### PR TITLE
CIAC-16176: Add fetch-assets to CyberArk Identity Security Platform

### PR DESCRIPTION
**Pull request by amshamah419**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-dc.paloaltonetworks.com/browse/CIAC-16176

## Description
Adds snapshot-based collection of CyberArk Cloud Directory data (Users, Groups, Roles, Applications) to the existing CyberArkISP integration via the new fetch-assets / isFetchAssets pipeline.

Adds four narrow debug commands:
- cyberark-isp-get-users
- cyberark-isp-get-groups
- cyberark-isp-get-roles
- cyberark-isp-get-applications

Tests: 17 new unit tests covering the multi-select parser, Redrock response shape extraction, single-page seal, multi-page pagination with snapshot_id reuse, multi-source per-invocation work bound, partial failure isolation, empty-source seal, and the manual debug command push/no-push branches.

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by marketplace-ai-reviewer** [comment]
🤖 Analysis started. Please wait for results...

File: Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.py [hunk 1]
```diff
@@ -463,6 +596,208 @@ def get_stream_results(self, cursor_ref: str) -> tuple[list[dict[str, Any]], str
...
+
+# region Redrock client (Directory Data / fetch-assets)
+# =================================
+# Separate HTTP client for the new Cloud Directory Snapshots feature.
+# Lives alongside Client (audit-events) so the two flows can evolve
+# independently. See CIAC-16176 for the design discussion.
+# =================================
+
+
+class RedrockClient(BaseClient):
```
Comment (marketplace-ai-reviewer) [inline Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.py:609]: New client classes must inherit from `ContentClient` instead of `BaseClient`.

File: Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.py [hunk 2]
```diff
@@ -639,6 +974,294 @@ def fetch_events_command(client: Client) -> None:
...
+            redrock_client=redrock_client,
+            source=current_source,
+            page_number=current_page,
+            page_size=page_size,
+        )
+    except Exception as error:
+        # Q4 partial-failure: drop this source from the cycle so the others
+        # can still complete. State for the failed source is reset so it
+        # restarts cleanly on the next cycle.
+        demisto.error(
```
Comment (marketplace-ai-reviewer) [inline Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.py:1149]: Include `traceback.format_exc()` in the error log for comprehensive debugging.

File: Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.py [hunk 3]
```diff
@@ -639,6 +974,294 @@ def fetch_events_command(client: Client) -> None:
...
+        snapshot_id = str(round(time.time() * 1000))
+        push_snapshot(annotated, source, snapshot_id=snapshot_id, items_count=len(annotated))
+        msg = (
+            f"Successfully retrieved and pushed {len(annotated)} {source.value} record(s) to "
+            f"{DIRECTORY_VENDOR}_{PRODUCT_BY_SOURCE[source]}_raw (snapshot_id={snapshot_id})."
+        )
+        return msg
+
+    readable = tableToMarkdown(f"{INTEGRATION_NAME} - {source.value}", annotated, removeNull=True)
+    return CommandResults(
```
Comment (marketplace-ai-reviewer) [inline Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.py:1282]: Include `raw_response` in `CommandResults`.

File: Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.py [hunk 4]
```diff
@@ -463,6 +596,208 @@ def get_stream_results(self, cursor_ref: str) -> tuple[list[dict[str, Any]], str
...
+        row = entry.get("Row") if isinstance(entry, dict) else None
+        if isinstance(row, dict):
+            rows.append(row)
+        elif isinstance(entry, dict):
+            # Some sources may return rows directly without a Row wrapper.
+            rows.append(entry)
+
+    full_count = result_block.get("FullCount")
+    page_count = result_block.get("Count", len(rows))
+    has_more = bool(full_count and isinstance(full_count, int) and page_count and page_count >= 0 and full_count > page_count)
```
Comment (marketplace-ai-reviewer) [inline Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.py]: The `has_more` calculation incorrectly compares `FullCount` (total items) with `page_count` (items in the current page), leading to an unnecessary API call.

File: Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP_test.py [hunk 5]
```diff
@@ -1380,3 +1380,359 @@ def test_main_parse_params_error(mocker, capfd):
...
+    send_mock = mocker.patch("CyberArkISP.send_data_to_xsiam")
+
+    config = _build_fetch_assets_config([DirectorySource.USERS])
+    fetch_assets_command(redrock_client, config)
+
+    # snapshot_id is reused on the send and persisted in last-run.
+    assert send_mock.call_args.kwargs["snapshot_id"] == "snap-abc-123"
+    payload = set_last_run_mock.call_args[0][0]
+    assert payload["snapshot_id"] == "snap-abc-123"
+    # cumulative total now 3+2 = 5; that's the seal because Count=2 == FullCount=5? no, FullCount=5 > Count=2 → still has_more
```
Comment (marketplace-ai-reviewer) [inline Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP_test.py:1611]: This test codifies the `has_more` bug mentioned in `CyberArkISP.py`.

File: Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.yml [hunk 6]
```diff
@@ -71,8 +72,136 @@ configuration:
...
+  display: Maximum number of records per Redrock page
+  required: false
+  type: 0
+  section: Assets
+  advanced: true
+  defaultvalue: "10000"
+  additionalinfo: Maximum number of records to request per page from the Redrock API (default 10000). Snapshots are paginated automatically across multiple invocations within a fetch cycle when needed.
 script:
   commands:
+  - name: cyberark-isp-get-users
```
Comment (marketplace-ai-reviewer) [inline Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.yml:105]: Command names should follow the `vendor-entity-action` pattern. Consider renaming `cyberark-isp-get-users` to `cyberark-isp-user-get`.

File: Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.yml [hunk 7]
```diff
@@ -71,8 +72,136 @@ configuration:
...
+    - contextPath: CyberArkISP.User.DisplayName
+      description: Display name of the user.
+      type: String
+    - contextPath: CyberArkISP.User.Email
+      description: Email address of the user.
+      type: String
+    - contextPath: CyberArkISP.User.LastLogin
+      description: Timestamp of the user's last login.
+      type: Date
+  - name: cyberark-isp-get-groups
```
Comment (marketplace-ai-reviewer) [inline Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.yml:137]: Command names should follow the `vendor-entity-action` pattern. Consider renaming `cyberark-isp-get-groups` to `cyberark-isp-group-get`.

File: Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.yml [hunk 8]
```diff
@@ -71,8 +72,136 @@ configuration:
...
+      - "true"
+      - "false"
+    outputs:
+    - contextPath: CyberArkISP.Group.ID
+      description: Unique identifier of the group.
+      type: String
+    - contextPath: CyberArkISP.Group.Name
+      description: Name of the group.
+      type: String
+  - name: cyberark-isp-get-roles
```
Comment (marketplace-ai-reviewer) [inline Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.yml:158]: Command names should follow the `vendor-entity-action` pattern. Consider renaming `cyberark-isp-get-roles` to `cyberark-isp-role-get`.

File: Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.yml [hunk 9]
```diff
@@ -71,8 +72,136 @@ configuration:
...
+    - contextPath: CyberArkISP.Role.ID
+      description: Unique identifier of the role.
+      type: String
+    - contextPath: CyberArkISP.Role.Name
+      description: Name of the role.
+      type: String
+    - contextPath: CyberArkISP.Role.Description
+      description: Description of the role.
+      type: String
+  - name: cyberark-isp-get-applications
```
Comment (marketplace-ai-reviewer) [inline Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.yml:182]: Command names should follow the `vendor-entity-action` pattern. Consider renaming `cyberark-isp-get-applications` to `cyberark-isp-application-get`.

File: Packs/CyberArkPAS/Integrations/CyberArkISP/README.md [hunk 10]
```diff
@@ -69,17 +69,22 @@ Before configuring the integration, you must complete the following prerequisite
...
+| Audit Server URL | The Audit API base URL from the SIEM integration \(e.g., https://example-domain.audit.cyberark.cloud\). | True |
+| Identity URL | The CyberArk Identity FQDN for OAuth2 authentication \(e.g., https://abc1234.id.cyberark.cloud\). | True |
+| Web App ID | The Application ID of the OAuth2 Server web app configured in Identity Administration. | True |
 | API Key | The API key from the SIEM integration created in the Administration space. | True |
-| Trust any certificate (not secure) | When selected, the integration will not verify SSL certificates. | False |
-| Use system proxy settings | When selected, the integration will use the system proxy settings. | False |
-| Maximum number of audit events per fetch | Maximum number of events to fetch per collection cycle.<br/>Default: 10000<br/>Note: The API returns a maximum of 1000 events per page. | False |
+| Client ID | Service username without domain \(configured as OAuth confidential client\). | True |
+| Client Secret | Service user password for OAuth2 authentication. | True |
+| Trust any certificate (not secure) |  | False |
```
Comment (marketplace-ai-reviewer) [inline Packs/CyberArkPAS/Integrations/CyberArkISP/README.md]: The description for this parameter is missing in the generated table.

```suggestion
| Trust any certificate (not secure) | When selected, the integration will not verify SSL certificates. | False |
```
To ensure this description is preserved during future documentation generation, please add it to the `additionalinfo` field of the parameter in the integration YAML file.

File: Packs/CyberArkPAS/Integrations/CyberArkISP/README.md [hunk 11]
```diff
@@ -69,17 +69,22 @@ Before configuring the integration, you must complete the following prerequisite
...
+| Use system proxy settings |  | False |
```
Comment (marketplace-ai-reviewer) [inline Packs/CyberArkPAS/Integrations/CyberArkISP/README.md]: The description for this parameter is missing in the generated table.

```suggestion
| Use system proxy settings | When selected, the integration will use the system proxy settings. | False |
```
To ensure this description is preserved during future documentation generation, please add it to the `additionalinfo` field of the parameter in the integration YAML file.

File: Packs/CyberArkPAS/Integrations/CyberArkISP/README.md [hunk 12]
```diff
@@ -150,3 +155,74 @@ For more information about CyberArk Identity Security Platform integration, refe
...
+| limit | Maximum number of role records to retrieve. Default is 50. | Optional | 
+| should_push_assets | Set to true to push the retrieved roles to XSIAM as a snapshot in the cyberark_roles_raw dataset. Possible values are: true, false. Default is false. | Optional | 
+
+#### Context Output
+
+| **Path** | **Type** | **Description** |
+| --- | --- | --- |
+| CyberArkISP.Role.ID | String | Unique identifier of the role. | 
+| CyberArkISP.Role.Name | String | Name of the role. | 
+| CyberArkISP.Role.Description | String | Description of the role. |
```
Comment (marketplace-ai-reviewer) [inline Packs/CyberArkPAS/Integrations/CyberArkISP/README.md]: The standard `Command example`, `Context Example`, and `Human Readable Output` sections are missing for this command (and other newly added commands).

```suggestion
| CyberArkISP.Role.Description | String | Description of the role. |

#### Command example
`!cyberark-isp-get-roles limit=50`

#### Context Example

#### Human Readable Output
```
Please run `demisto-sdk generate-docs --insecure -e <path/to/command_examples.txt>` to regenerate the entire command section with proper formatting, examples, and context output. For more details, refer to the [Creating Documentation](https://xsoar.pan.dev/docs/documentation/readme_file#creating-documentation) guidelines.

**Comment by marketplace-ai-reviewer** [comment]
<details>
<summary><strong>🤖 AI Review Disclaimer</strong></summary>

This review was generated by an AI-powered tool and may contain inaccuracies. Please be advised, and we extend our sincere apologies for any inconvenience this may cause.
</details>

**Review by marketplace-ai-reviewer — COMMENTED** [review]
Hi! Thanks for your contribution to the CyberArkPAS pack. I've reviewed the PR and noted a few adjustments needed for standard compliance and logic:

* **Code & Logic**: Update the client class to inherit from `ContentClient` and fix the `has_more` pagination calculation (along with its corresponding test).
* **Naming Standards**: Rename the newly added commands to follow the standard `vendor-entity-action` format (e.g., `cyberark-isp-user-get`).
* **Documentation**: Add the missing parameter descriptions to the YAML and regenerate the README using `demisto-sdk generate-docs` to fix structural issues.

Thanks again for your hard work on this!

Additionally, please address the following file-level notes:
- `Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.yml`: - The first key defined at the root level of the `.yml` file must always be `name` or `display`.
- The `cyberark-isp-get-users` command is missing from the `README.md` documentation.
- `Packs/CyberArkPAS/Integrations/CyberArkISP/README.md`: The README appears to have been manually edited rather than fully generated using `demisto-sdk generate-docs`. Several structural issues were found:

1. The new commands (`cyberark-isp-get-roles`, `cyberark-isp-get-applications`, `cyberark-isp-get-groups`) are appended at the very end of the file, after the `## Additional Resources` section, instead of under `## Commands`.
2. The `cyberark-isp-get-users` command is completely missing from the README, despite being added to the integration YAML.
3. The standard `Command example`, `Context Example`, and `Human Readable Output` sections are missing for all the newly added commands.

Please regenerate the README using `demisto-sdk generate-docs --insecure -e <path/to/command_examples.txt>` to ensure all commands are correctly documented, properly formatted, and placed in the correct sections. For more details, refer to the [Creating Documentation](https://xsoar.pan.dev/docs/documentation/readme_file#creating-documentation) guidelines.
- `Packs/CyberArkPAS/pack_metadata.json`: Consider adding the `githubUser` field with a list of valid GitHub usernames.

 @DanielTal87, @amshamah419 please review and approve the results generated by the AI Reviewer by responding 👍 on this comment.

**Comment by marketplace-ai-reviewer** [comment]
🤖 Analysis started. Please wait for results...

File: Packs/CyberArkPAS/Integrations/CyberArkISP/README.md [hunk 13]
```diff
@@ -150,3 +153,74 @@ For more information about CyberArk Identity Security Platform integration, refe
...
+#### Base Command
+
+`cyberark-isp-get-roles`
+
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| limit | Maximum number of role records to retrieve. Default is 50. | Optional |
+| should_push_assets | Set to true to push the retrieved roles to XSIAM as a snapshot in the cyberark_roles_raw dataset. Possible values are: true, false. Default is false. | Optional |
```
Comment (marketplace-ai-reviewer) [inline Packs/CyberArkPAS/Integrations/CyberArkISP/README.md:197]: According to the Cortex Branding Guidelines, when XSIAM is mentioned, the product name must include the "Cortex" prefix. Please update "XSIAM" to "Cortex XSIAM" here and in the descriptions for `cyberark-isp-get-applications` (line 195) and `cyberark-isp-get-groups` (line 219).

```suggestion
| should_push_assets | Set to true to push the retrieved roles to Cortex XSIAM as a snapshot in the cyberark_roles_raw dataset. Possible values are: true, false. Default is false. | Optional |
```

File: Packs/CyberArkPAS/Integrations/CyberArkISP/README.md [hunk 14]
```diff
@@ -150,3 +153,74 @@ For more information about CyberArk Identity Security Platform integration, refe
...
+
+***
+Manually fetches CyberArk Cloud Directory role snapshots via the Redrock Query API. For developing/debugging only.
+
+#### Base Command
+
+`cyberark-isp-get-roles`
+
+#### Input
+
```
Comment (marketplace-ai-reviewer) [inline Packs/CyberArkPAS/Integrations/CyberArkISP/README.md:193]: Consider using a complete sentence for clarity. Please apply this change to the other newly added commands as well (lines 191, 215).

```suggestion
Manually fetches CyberArk Cloud Directory role snapshots via the Redrock Query API. This command is for development and debugging purposes only.
```

**Comment by marketplace-ai-reviewer** [comment]
<details>
<summary><strong>🤖 AI Review Disclaimer</strong></summary>

This review was generated by an AI-powered tool and may contain inaccuracies. Please be advised, and we extend our sincere apologies for any inconvenience this may cause.
</details>

**Review by marketplace-ai-reviewer — COMMENTED** [review]
#### 🔄 Re-reviewing updates...
> [!IMPORTANT]
> To maintain review integrity, threads marked as **Resolved** without accompanying changes or explanations will be automatically **unresolved** for further verification.

Thanks for your contribution to the CyberArk ISP integration! I've reviewed the updates and left a few minor documentation notes. The main adjustments needed are ensuring proper Cortex branding by using "Cortex XSIAM" and tweaking a few command descriptions to use complete sentences for better readability. Great work overall!

 @marketplace-ai-reviewer, @DanielTal87, @amshamah419 please review and approve the results generated by the AI Reviewer by responding 👍 on this comment.

**Comment by amshamah419** [comment]
## PR Review Response — `marketplace-ai-reviewer`

Processed 12 inline review comments. Latest commit: `2d9d611ac75`.

### Summary

| Action | Count |
|--------|-------|
| ✅ Implemented | 7 |
| 💡 Justified | 5 |

### Per-comment dispositions

| # | Comment (summary) | File | Action |
|---|---|---|---|
| 1 | `RedrockClient` should inherit from `ContentClient` instead of `BaseClient` | `CyberArkISP.py` | 💡 Justify |
| 2 | Include `traceback.format_exc()` in fetch-assets error log | `CyberArkISP.py:1144` | ✅ Implement |
| 3 | Pass `raw_response` to `CommandResults` in `get_assets_command` | `CyberArkISP.py:1282` | ✅ Implement |
| 4 | Clarify/tighten `has_more` calculation in `extract_rows_from_redrock_response` | `CyberArkISP.py:757-810` | ✅ Implement |
| 5 | Add test for the last-partial-page `has_more` edge case | `CyberArkISP_test.py:1472` | ✅ Implement |
| 6-9 | Rename `cyberark-isp-get-{users,groups,roles,applications}` → `cyberark-isp-{...}-get` | `CyberArkISP.yml` | 💡 Justify (×4) |
| 10 | README: `Trust any certificate` param missing description | `CyberArkISP.yml` | ✅ Implement |
| 11 | README: `Use system proxy settings` param missing description | `CyberArkISP.yml` | ✅ Implement |
| 12 | README: missing Command / Context / Human-Readable examples | `command_examples.txt` + `README.md` | ✅ Implement |

### Justifications

**#1 — RedrockClient inherits from `BaseClient`**
The pre-existing `Client` class in this same file (lines 401-596) also inherits from `BaseClient`, not `ContentClient`. Switching only `RedrockClient` would introduce inconsistency within the integration. `BaseClient` already provides `_http_request` with retries, error handling, and proxy/SSL handling — exactly what `RedrockClient` needs. If the broader repository moves to `ContentClient`, both clients should migrate together in a dedicated refactor PR.

**#6-9 — Command naming `cyberark-isp-get-X` vs `cyberark-isp-X-get`**
The existing public command in this integration is `cyberark-isp-get-events`, established and consumed by playbooks. Renaming the new commands to `cyberark-isp-{users,groups,roles,applications}-get` would break intra-integration consistency. Following the established `<vendor>-<product>-get-<entity>` pattern keeps a uniform command grammar for users of this integration.

### Implementation details

**#2 — Traceback in error log** ([`CyberArkISP.py:1144`](Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.py:1144))
```python
demisto.error(
    f"[fetch_assets] Source {source.value} failed: {exc}. "
    f"Skipping source for this cycle. Traceback:\n{traceback.format_exc()}"
)
```

**#3 — `raw_response` on `CommandResults`** ([`CyberArkISP.py:1282`](Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.py:1282))
```python
return CommandResults(
    outputs_prefix=f"CyberArkISP.{CONTEXT_PREFIX_BY_SOURCE[source]}",
    outputs_key_field="ID",
    outputs=annotated,
    raw_response=annotated,
    readable_output=tableToMarkdown(...),
)
```

**#4 — `has_more` clarifying comment** ([`CyberArkISP.py:757-810`](Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.py:757))
Added inline comment explaining that `has_more` is `True` when either (a) `FullCount > rows_seen + len(rows)` per server, OR (b) the page is full (`len(rows) >= page_size`) — the latter handles the edge case where the server doesn't return a reliable `FullCount` for the last partial page.

**#5 — Regression test** ([`CyberArkISP_test.py:1472`](Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP_test.py:1472))
New test `test_extract_rows_from_redrock_response_last_partial_page_signals_more` documents the page-full → `has_more=True` behavior.

**#10/#11 — Param `additionalinfo`** ([`CyberArkISP.yml`](Packs/CyberArkPAS/Integrations/CyberArkISP/CyberArkISP.yml))
Added `additionalinfo` to `insecure` ("Trust any certificate (not secure). Use only for testing or with self-signed certificates.") and `proxy` ("Use the HTTP/HTTPS proxy configured in the engine's environment variables.").

**#12 — Command examples** ([`command_examples.txt`](Packs/CyberArkPAS/Integrations/CyberArkISP/command_examples.txt))
Created `command_examples.txt` with 5 example invocations covering `cyberark-isp-get-events`, `cyberark-isp-get-users`, `cyberark-isp-get-groups`, `cyberark-isp-get-roles`, and `cyberark-isp-get-applications`. Regenerated `README.md` via `demisto-sdk generate-docs` so each command now includes the Command / Context Example / Human-Readable Output blocks.

### Pre-commit verification

All code-level hooks pass on the latest commit:

| Hook | Result |
|---|---|
| `ruff-format` | ✅ Passed |
| `ruff-py3.12` | ✅ Passed |
| `pylint-in-docker` | ✅ Passed |
| `mypy-in-docker` | ✅ Passed |
| `pytest-in-docker` | ✅ Passed |
| `xsoar-lint` | ✅ Passed |
| `Validate README.md (MDX)` | ✅ Passed |
| `markdownlint-cli2` | ✅ Passed |
| `secrets` | ✅ Passed |
| `pycln` | ⚠️ Environmental (`.DS_Store` in scan path; not a code issue) |

Ready for re-review.

**Review by DanielTal87 — APPROVED** [review]
